src/Umbraco.Cms.ManagementApi/Controllers/RecycleBinControllerBase.cs

```diff
@@ -0,0 +1,115 @@
+    private readonly string _itemUdiType;
+
+    protected RecycleBinControllerBase(IEntityService entityService)
+    {
+        _entityService = entityService;
+        // ReSharper disable once VirtualMemberCallInConstructor
+        _itemUdiType = ItemObjectType.GetUdiType();
+    }
+
+    protected abstract UmbracoObjectTypes ItemObjectType { get; }
+
+    protected async Task<ActionResult<PagedResult<TItem>>> GetRoot(long pageNumber, int pageSize)
+    {
+        IEntitySlim[] rootEntities = GetPagedRootEntities(pageNumber, pageSize, out var totalItems);
+
+        TItem[] treeItemViewModels = MapRecycleBinViewModels(null, rootEntities);
+
+        PagedResult<TItem> result = PagedResult(treeItemViewModels, pageNumber, pageSize, totalItems);
+        return await Task.FromResult(Ok(result));
+    }
+
+    protected async Task<ActionResult<PagedResult<TItem>>> GetChildren(Guid parentKey, long pageNumber, int pageSize)
+    {
+        IEntitySlim[] children = GetPagedChildEntities(parentKey, pageNumber, pageSize, out var totalItems);
+
+        TItem[] treeItemViewModels = MapRecycleBinViewModels(parentKey, children);
+
+        PagedResult<TItem> result = PagedResult(treeItemViewModels, pageNumber, pageSize, totalItems);
+
+        return await Task.FromResult(Ok(result));
+    }
+
+    protected virtual TItem MapRecycleBinViewModel(Guid? parentKey, IEntitySlim entity)
+    {
+        if (entity == null)
+        {
+            throw new ArgumentNullException(nameof(entity));
+        }
+
+        var viewModel = new TItem
+        {
+            Icon = _itemUdiType,
+            Name = entity.Name!,
+            Key = entity.Key,
+            Type = _itemUdiType,
+            HasChildren = entity.HasChildren,
+            IsContainer = entity.IsContainer,
+            ParentKey = parentKey
+        };
+
+        return viewModel;
+    }
+
+    private IEntitySlim[] GetPagedRootEntities(long pageNumber, int pageSize, out long totalItems)
+    {
+        // TODO: EntityService needs to be expanded to get root identifier (key?) by UmbracoObjectType - for now this is a workaround
+        int rootId = ItemObjectType == UmbracoObjectTypes.Document
+            ? Constants.System.RecycleBinContent
+            : ItemObjectType == UmbracoObjectTypes.Media
+                ? Constants.System.RecycleBinMedia
+                : throw new ArgumentOutOfRangeException(nameof(ItemObjectType), "Only Document and Media are supported");
+
+        // TODO: EntityService needs to be able to return paged trashed children - for now we'll get all children
+        // IEntitySlim[] rootEntities = EntityService
+        //     .GetPagedChildren(rootId, ItemObjectType, pageNumber, pageSize, out totalItems)
+        //     .ToArray();
+        IEntitySlim[] rootEntities = _entityService
+            .GetChildren(rootId, ItemObjectType)
+            .ToArray();
+        totalItems = rootEntities.Length;
+
+        return rootEntities;
+    }
```

Comment:
Not necessarily a huge thing, but I can't help wondering if this should be the responsibility of the controller, and not something like for instance a `IRecycleBinEntityService` or something similar

Reply:
This is definitively not the responsibility of this controller; there will be tasks to amend `EntityService` to cater for the needs of the new backoffice; we will circle back and remove this ill-placed logic later.